### ISSUE TITLE
Preconstruct memory heap failed

### ISSUE THREAD

**Issue by apuntovanini** [preconstruct/preconstruct#314]
Hi, we're facing a weird issue about memory consumption. We build indeed many packages altogether with preconstruct build, but memory usage keeps increasing (seems something related to garbage collection or such).
When we increase node's to a huge 16384 --max-old-space-size=16384 we manage to build it in dev, not on travis.
Do you think it may be something related to rollup itself or to preconstruct?
It's really difficult for me to reproduce, let's say we're trying to use preconstruct with a repo similar in size and structure to what Atlaskit-mk2 was before refactoring to Atlassian design system

You can find it at https://github.com/uidu-org/guidu

**Comment by mitchellhamilton** [comment]
I just published `@preconstruct/cli@1.1.27` which should fix the memory problem though the build will likely be much faster if you remove the tsconfigs in all the packages and instead only have one at the root. Having a tsconfig per package means that TS has to do much more unnecessary type checking(e.g. if a bunch of packages have a dependency on another package and you have one tsconfig per package, TS has to type check the dependency a number of times equal to the number of packages that depend on it vs just once if there's only one tsconfig)

**Comment by sarahbethfederman** [comment]
Oh epic! Has there been a change with the way tsconfigs are handled? I originally had one at the root and I received errors about it and had to copy it to each package

**Comment by mitchellhamilton** [comment]
No change, Preconstruct has always supported having just one tsconfig at the root or one per package, the only change was fixing a memory leak when using one per package(having one tsconfig at the root is definitely the recommended approach but Preconstruct doesn't enforce it).

(There was a bug fixed recently where the tsconfig wasn't actually being used which resulted in some incorrect type declarations being generated - https://github.com/preconstruct/preconstruct/commit/6b3f95e2db4704722dc5d3c9ade64fb2d3a76965)

> I originally had one at the root and I received errors about it and had to copy it to each package

That's v strange

**Comment by sarahbethfederman** [comment]
Hm, let me try it again and see what happens. I was having a pretty clear error for each package that said no tsconfig was found. I ended up making one per package that extended the root one.

**Comment by sarahbethfederman** [comment]
Yeah still having errors after deleting the tsconfigs in packages. My tsconfig in the root looks like this: 
```
{
  "extends": "./node_modules/@lendi/typescript-config/tsconfig",
  "compilerOptions": {
    "lib": ["dom", "es2018"],
    "jsx": "react",
    "module": "commonjs",
    "target": "es5",
    "esModuleInterop": true,
    "isolatedModules": true, // this is just for clarity, babel assumes this https://github.com/babel/babel/issues/11464#issuecomment-617609332
    "baseUrl": "./",
    "paths": {
      "@lendi-ui/*-utils": ["./utils/*-utils/src"],
      "@lendi-ui/*": ["./packages/*/src"]
    }
  },
  "include": ["typings/**/*.d.ts", "@(packages|utils)/*/src/**/*.ts", "@(packages|utils)/*/src/**/*.tsx"],
  "exclude": [
    "node_modules",
    "@(packages|utils)/*/src/**/*.d.ts",
    "@(packages|utils)/*/src/**/*.test.ts",
    "@(packages|utils)/*/src/**/*.test.tsx",
    "@(packages|utils)/*/src/packages/*/**/*.example.tsx",
    "@(packages|utils)/*/src/examples/**.tsx",
    "@(packages|utils)/*/src/**/*.doc.ts",
    "@(packages|utils)/*/src/**/*.doc.tsx"
  ]
}
```
and I'm getting errors that look like this: 
```
$ preconstruct build
🎁 info building bundles!
🎁 error @lendi-ui/accordion Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/accordion/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/address-picker Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/address-picker/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/alert Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/alert/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/auth-loader Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/auth-loader/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/auto-complete Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/auto-complete/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/breakpoint Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/breakpoint/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/button Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/button/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/card Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/card/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/carousel Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/carousel/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/chat-bubble Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/chat-bubble/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/checkbox Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/checkbox/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/collapse Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/collapse/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/color Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/color/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/container Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/container/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/date-picker Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/date-picker/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/depth Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/depth/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/dropdown Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/dropdown/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/fancy-icon Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/fancy-icon/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/feature-panel Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/feature-panel/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/field Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/field/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/footer Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/footer/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/grid Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/grid/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/icon Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/icon/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/legal-stuff Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/legal-stuff/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/lender-logos Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/lender-logos/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/logo Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/logo/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/menu-dropdown Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/menu-dropdown/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/modal Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/modal/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/navbar-base Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/navbar-base/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/navbar-vertical Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/navbar-vertical/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/navigation Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/navigation/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/overlay Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/overlay/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/pagination Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/pagination/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/popover Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/popover/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/radio Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/radio/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/sidebar Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/sidebar/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/spacing Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/spacing/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/spinner Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/spinner/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/status-badge Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/status-badge/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/tabs Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/tabs/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/text-area Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/text-area/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/text-bubble Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/text-bubble/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/text-input Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/text-input/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/theme-domain Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/theme-domain/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/theme Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/theme/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/toggle-switch Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/toggle-switch/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/tooltip Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/tooltip/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/transition Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/transition/src/index.tsx is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/typography Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/typography/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 error @lendi-ui/utils Could not generate type declarations because /Users/sarahfederman/Documents/lui/packages/utils/src/index.ts is not in a TypeScript project. Make sure this file is included in your tsconfig.
🎁 info If want to learn more about the above error, check https://preconstruct.tools/errors
🎁 info If the error is not there and you want to learn more about it, open an issue at https://github.com/preconstruct/preconstruct/issues/new
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
sarahfederman@syd-sfederman-l2 lui % 
```

**Comment by mitchellhamilton** [comment]
I think the `@`s in your tsconfig are the issue. The packages are at `packages/...` but in your tsconfig, you're saying that they're at `@packages/...`.

Also, you shouldn't need to use the `paths` option if you're using `preconstruct dev`

**Comment by sarahbethfederman** [comment]
Oh weird, it didn't like that glob pattern at all. 

Holy crap our build is now less than a minute! 😱 Thanks friend!!

**Comment by apuntovanini** [comment]
Impressive @mitchellhamilton, not only it does work now without too much pressure on memory, but it builds much much faster! 😱
Only things that remains for us is the "unhandled Rollup error" with @loadable/components, should I open a new issue about this? I think these are unrelated

**Comment by Andarist** [comment]
Just a note - while the single root `tsconfig.json` works fine most of the time, it might be sometimes required to use multiple when targeting different environments. The perf issue for such scenarios should be mitigated by using project references though but when I've tried to use them with preconstruct in the past I've failed - it wasn't as important for me back then so I haven't investigated the issue further and just switched to a root `tsconfig.json`. Nevertheless, I think it would be worthwhile to make sure that this works with project references.
